File: baekjoon/7569_토마토/B_7569_2.js
const Node = function(value, next = null) {
    this.value = value
    this.next = next
}

const Queue = function() {
    this.head = new Node(null)
    this.tail = this.head
}

Queue.prototype.isEmpty = function() {
    if (this.head.value === null && this.tail.value === null) {
        return true
    }

    return false
}

Queue.prototype.enQueue = function(item) {
    if (this.isEmpty()) {
        this.head = new Node(item)
        this.tail = this.head
    } else {
        const tail = this.tail
        this.tail = new Node(item)
        tail.next = this.tail
    }
}

Queue.prototype.deQueue = function() {
    if (this.isEmpty() === false) {
        const popNumber = this.head.value

        if (this.head === this.tail) {
            this.head = new Node(null)
            this.tail = this.head
        } else {
            this.head = this.head.next
        }

        return popNumber
    }
}

const isAllRipen = function() {
    for (let i = 0; i < H; i++) {
        for (let j = 0; j < N; j++) {
            for (let k = 0; k < M; k++) {
                if (boxes[i][j][k] === 0) {
                    return false
                }
            }
        }
    }

    return true
}

const isValidRange = function(h, n, m) {
    if (0 <= h && h < H && 0 <= n && n < N && 0 <= m && m < M) {
        return true
    }

    return false
}

const bfs = function(q) {
    let maxDay = 1

    const zDirection = [-1, 1, 0, 0, 0, 0]
    const yDirection = [0, 0, -1, 0, 1, 0]
    const xDirection = [0, 0, 0, 1, 0, -1]

    while (q.isEmpty() === false) {
        const [z, y, x] = q.deQueue()

        for (let i = 0; i < 6; i++) {
            const zIdx = z + zDirection[i]
            const yIdx = y + yDirection[i]
            const xIdx = x + xDirection[i]

            if (isValidRange(zIdx, yIdx, xIdx)) {
                if (boxes[zIdx][yIdx][xIdx] === 0) {
                    boxes[zIdx][yIdx][xIdx] = boxes[z][y][x] + 1
                    maxDay = Math.max(maxDay, boxes[zIdx][yIdx][xIdx])
                    q.enQueue([zIdx, yIdx, xIdx])
                }
            }
        }
    }

    return maxDay
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const [M, N, H] = input[0].split(' ').map(v => parseInt(v))
let boxes = []

let idx = 1
for (let i = 0; i < H; i++) {
    let box = []
    for (let j = idx; j < idx + N; j++) {
        const oneBox = input[j].split(' ').map(v => parseInt(v))
        box.push(oneBox)
    }
    boxes.push(box)
    idx += N
}

const deq = new Queue()

for (let i = 0; i < H; i++) {
    for (let j = 0; j < N; j++) {
        for (let k = 0; k < M; k++) {
            if (boxes[i][j][k] === 1) {
                deq.enQueue([i, j, k])
                
            }
        }
    }
}

const answer = bfs(deq)

if (isAllRipen()) {
    console.log(answer - 1)
} else {
    console.log(-1)
}

File: baekjoon/7569_토마토/B_7569_2.py
from collections import deque
import sys


def find_zero():
    for a in range(H):
        for b in range(N):
            for c in range(M):
                if boxes[a][b][c] == 0:
                    return 1
    return 0


def bfs(locations):
    deq = deque()
    for location in locations:
        deq.append(location)

    # 위, 아래, 앞, 오른, 뒤, 왼
    z_location = [-1, 1, 0, 0, 0, 0]
    y_location = [0, 0, -1, 0, 1, 0]
    x_location = [0, 0, 0, 1, 0, -1]

    # boxes 에 0이 없는 경우도 있다
    # 이때는 while 문이 아예 돌지 않기 때문에 바로 0을 출력할 수 있도록
    # max_days 를 0으로 설정한다
    max_days = 0
    while deq:
        z, y, x, cnt = deq.popleft()
        for m in range(6):
            z_idx = z_location[m] + z
            y_idx = y_location[m] + y
            x_idx = x_location[m] + x
            if 0 <= z_idx < H and 0 <= y_idx < N and 0 <= x_idx < M:
                if boxes[z_idx][y_idx][x_idx] == 0:
                    boxes[z_idx][y_idx][x_idx] = 1
                    deq.append([z_idx, y_idx, x_idx, cnt + 1])
                    max_days = max(max_days, cnt + 1)

    if not find_zero():
        return max_days
    return -1


# 가로, 세로, 상자수
M, N, H = map(int, sys.stdin.readline().split())
one = []
boxes = []
for i in range(H):
    box = []
    for j in range(N):
        a_line_of_box = list(map(int, sys.stdin.readline().split()))
        box.append(a_line_of_box)
        for k in range(M):
            if a_line_of_box[k] == 1:
                one.append([i, j, k, 0])
    boxes.append(box)

answer = bfs(one)
sys.stdout.write(str(answer))
